src/ray/core_worker/store_provider/memory_store/memory_store.h

```diff
@@ -0,0 +1,94 @@
+#ifndef RAY_CORE_WORKER_MEMORY_STORE_H
+#define RAY_CORE_WORKER_MEMORY_STORE_H
+
+#include "ray/common/buffer.h"
+#include "ray/common/id.h"
+#include "ray/common/status.h"
+#include "ray/core_worker/common.h"
+
+namespace ray {
+
+class ReferencedRayObject;
+class GetOrWaitRequest;
+class CoreWorkerMemoryStore;
+
+/// A class that represents an object in memory store.
+class ObjectEntry {
+ public:
+  ObjectEntry(const ObjectID &object_id, const RayObject &object);
+
+  std::shared_ptr<RayObject> GetObject() const { return object_; }
+  int IncreaseRefcnt() { return ++refcnt_; }
+  int DecreaseRefcnt() { return --refcnt_; }
+  int Refcnt() { return refcnt_; }
+
+  std::shared_ptr<ReferencedRayObject> CreateReferencedObject(
+      std::shared_ptr<CoreWorkerMemoryStore> provider);
+
+ private:
+  const ObjectID object_id_;
+  std::shared_ptr<RayObject> object_;
+  std::atomic<uint16_t> refcnt_;
+};
+
+/// This cache currently contains unconsumed objects, which are the objects
+/// that have been `put` but haven't been `get`.
+class EvictionCache {
+ public:
+  EvictionCache() {}
+
+  void Add(const ObjectID &key, uint64_t size);
+
+  void Remove(const ObjectID &key);
+
+  uint64_t ChooseObjectsToEvict(uint64_t num_bytes_required,
+                                std::vector<ObjectID> *objects_to_evict);
+
+ private:
+  /// A doubly-linked list containing the items in the cache.
+  typedef std::list<std::pair<ObjectID, uint64_t>> ItemList;
+  ItemList item_list_;
+  /// A hash table mapping the object ID of an object in the cache to its
+  /// location in the doubly linked list item_list_.
+  std::unordered_map<ObjectID, ItemList::iterator> item_map_;
+};
+
+/// The class provides implementations for local process memory store.
+/// An example usage for this is to retrieve the returned objects from direct
+/// actor call (see direct_actor_transport.cc).
+class CoreWorkerMemoryStore : public std::enable_shared_from_this<CoreWorkerMemoryStore> {
```

Comment:
This is not needed, instead, we should have a unique_ptr in the ObjectID, which will be updated to own the object once it has been transfered. This will be way simpler.

Reply:
Sorry I don't understand the suggestion on "have a unique_ptr in the ObjectID"? Could you explain a bit more on the suggested approach please? Thanks!

Reply:
Yes! Basically replace the ObjectID with something like this:

```
union ObjectIDData {
    std::unique_ptr<RayObject> object;
    uint8_t[kObjectIDSize] data;
}

class ObjectID {
 private:
   bool is_direct_object;
   ObjectIDData data;
}
```

And then if object_id.is_direct_object is set, the object data will be stored in object_id.data.object.

Do you understand what I mean?

Reply:
Thanks, yes I got your point now. Agreed that this would be much simpler.

But I'm also concerned about the ID size increase, as this would increase the ray ObjectID by another 8 bytes only for the memory store scenario.  And this would require a lot of changes in existing python/java/c++ code I think.

How about we instead use similar approach like existing `RayObject` in java, which basically wraps up object id and object data together. so as long it is in scope, it's guaranteed that the data is available. We can do something similar in python later. The changes should be limited comparing to the suggested approach.

https://github.com/ray-project/ray/blob/master/java/api/src/main/java/org/ray/api/RayObject.java
https://github.com/ray-project/ray/blob/master/java/runtime/src/main/java/org/ray/runtime/RayObjectImpl.java

How does this sound?

Reply:
And object IDs can be passed around. Consider that later we support the "object escape" scenario, when this object ID gets passed to another node, the unique_ptr would become invalid.

Reply:
The RayObject approach sounds good to me (it's equivalent to the ObjectID suggestion but cleaner), can you change the PR to do that?

Reply:
Right now python worker hasn't been migrated to the core worker yet. For this PR I can simplify the C++ changes by removing this object from memory store when user issues a `ray.get`, so that we don't the the `shared_from_this` stuff. When later python worker is ready, then a later PR can address it in python side without further changing C++ worker code.